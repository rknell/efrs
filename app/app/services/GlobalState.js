import { observable, observe, spy } from 'mobx'
import { Alert, AsyncStorage, Linking } from 'react-native'
import PushNotifications from './PushNotifications'
import Translate from '../services/Translate'

class GlobalState {
  // apiPath = 'http://192.168.1.8:1676/api'
  apiPath = 'https://www.efrs.io/api'
  // apiPath = 'http://localhost:1676/api'

  @observable model = {
    userToken: null,
    user: null,
    userPermissions: {
      permissions: {},
      groups: []
    },
    canPop: false,
    forms: [],
    cachedCalls: {
      get: {},
      post: []
    },
    isBusy: 0,
    pushToken: null
  }

  current = () => {
    return {
      user: this.model.user ? this.model.user._id : 'no user',
      group: this.model.userPermissions.groups.join(',')
    }
  }

  navigation = {
    resetTo: () => {
      console.log('RESETTING NAVIGATION')
    },
    push: () => {
      console.log('PUSHING NAVIGATION')
    }
  }

  saveSetting (key, value) {
    return AsyncStorage.setItem(`@EFRS.AppSettings:${key}`, value)
  }

  getSetting (key) {
    return AsyncStorage.getItem(`@EFRS.AppSettings:${key}`)
  }

  constructor () {
    spy(event => {
      if (event.type === 'action' || event.type === 'update') {
        // console.log(event);
      }
    })

    this.getSetting('formsData').then(data => {
      if (data) this.model.forms = JSON.parse(data)
    })

    const PNObj = new PushNotifications(this._setPushToken)
    this.pushNotifications = PNObj.messaging

    observe(this.model, 'user', () => {
      try {
        if (this.model.user && this.bugsnag) {
          this.bugsnag.setUser(this.model.user._id, `${this.model.user.firstName} ${this.model.user.lastName}`, this.model.user.email)
        }
      } catch (e) {
        console.error('could not set user for bugsnag', e)
      }
    })
  }

  _setPushToken = (token) => {
    this.model.pushToken = token
  }

  openUrl (url) {
    Linking.canOpenURL(url).then(supported => {
      if (!supported) {
        Alert.alert(Translate.do('Error'), Translate.do('Can\'t open this url on this device. Try installing an app that can handle this link type and try again.'))
      } else {
        return Linking.openURL(url)
      }
    }).catch(err => console.error('An error occurred', err))
  }
}

export default new GlobalState()
