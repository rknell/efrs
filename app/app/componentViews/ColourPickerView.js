import React, { Component } from 'react'
import { ScrollView, Text, TouchableOpacity, View } from 'react-native'
import { observer } from 'mobx-react/native'
import FieldBase from '../componentViews/FieldBase'
import NavBar from '../components/NavBar'
import GlobalState from '../services/GlobalState'
import tinycolor from 'tinycolor2'

@observer
class ColourPickerView extends Component {
  onPress = () => {
    GlobalState.navigation.push({ view: <ColourPickerSelectView {...this.props} /> })
  }

  render () {
    let settings = { Label: this.props.label }
    if (this.props.settings) settings = this.props.settings
    return (
      <FieldBase hidden={this.props.editMode ? false : settings['Hidden']} label={settings.Label} {...this.props}>
        <View style={{ flex: 1, justifyContent: 'center' }}>
          <TouchableOpacity style={{ flex: 1, justifyContent: 'center' }} disabled={this.props.readOnly} onPress={this.onPress}>
            {this.props.value ? <View style={{ backgroundColor: this.props.value, padding: 10 }} />
              : <Text style={{ color: 'dodgerblue' }}>Pick Colour</Text>
            }
          </TouchableOpacity>
        </View>
      </FieldBase>
    )
  }
}

export class ColourPickerSelectView extends Component {
  colours = [
    { name: 'Red', value: '#f34235' },
    { name: 'Pink', value: '#e81d62' },
    { name: 'Purple', value: '#9b26af' },
    { name: 'Indigo', value: '#3e50b4' },
    { name: 'Blue', value: '#2095f2' },
    { name: 'Light Blue', value: '#02a8f3' },
    { name: 'Cyan', value: '#00bbd3' },
    { name: 'Teal', value: '#009587' },
    { name: 'Green', value: '#4bae4f' },
    { name: 'Light Green', value: '#8ac249' },
    { name: 'Lime', value: '#ccdb38' },
    { name: 'Yellow', value: '#feea3a' },
    { name: 'Amber', value: '#fec006' },
    { name: 'Orange', value: '#fe9700' },
    { name: 'Deep Orange', value: '#fe5621' },
    { name: 'Brown', value: '#785447' },
    { name: 'Grey', value: '#9d9d9d' },
    { name: 'Blue Grey', value: '#5f7c8a' },
    { name: 'Black', value: '#000' },
    { name: 'White', value: '#fff' }
  ]

  onPress = (colour) => {
    this.props.onChange(colour.value)
    GlobalState.navigation.pop()
  }

  render () {
    return (
      <View style={{ flex: 1, backgroundColor: 'white' }}>
        <NavBar title='Select Colour' />
        <ScrollView style={{ flex: 1, backgroundColor: 'rgba(0,0,0,0.3)' }}
          contentContainerStyle={{ flexDirection: 'row', flexWrap: 'wrap', justifyContent: 'flex-start' }}>
          {this.colours.map(colour => <ColourBrick colour={colour} key={colour.value} onPress={this.onPress} />)}
        </ScrollView>
      </View>
    )
  }
}

export class ColourBrick extends Component {
  onPress = () => {
    this.props.onPress(this.props.colour)
  }

  render () {
    return (<TouchableOpacity key={this.props.colour.name} onPress={this.onPress}
      style={{ flexShrink: 0, flexGrow: 1, flexBasis: 100, height: 100, flex: 1 }}>
      <View style={{ padding: 10, flex: 1, backgroundColor: this.props.colour.value }}>
        <Text
          style={{ color: tinycolor(this.props.colour.value).isLight() ? 'black' : 'white' }}>{this.props.colour.name}</Text>
      </View>
    </TouchableOpacity>)
  }
}

export default ColourPickerView
