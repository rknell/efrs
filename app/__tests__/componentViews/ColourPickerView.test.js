import React from 'react'
import ColourPickerView, { ColourBrick, ColourPickerSelectView } from '../../app/componentViews/ColourPickerView'
import { shallow } from 'enzyme'
import GlobalState from '../../app/services/GlobalState'
import toJSON from 'enzyme-to-json'

it('should render the form control with no value selected', () => {
  const wrapper = shallow(<ColourPickerView settings={{ test: true }} />)
  expect(toJSON(wrapper)).toMatchSnapshot()
})

it('should render the form control with black selected', () => {
  const wrapper = shallow(<ColourPickerView value='black' />)
  expect(toJSON(wrapper)).toMatchSnapshot()
})

it('should render the form control in edit mode', () => {
  const wrapper = shallow(<ColourPickerView settings={{ test: true }} editMode />)
  expect(toJSON(wrapper)).toMatchSnapshot()
})

it('should navigate to the detail view when pressed', () => {
  let pushed = false
  GlobalState.navigation.push = () => { pushed = true }
  const wrapper = shallow(<ColourPickerView />)
  const instance = wrapper.instance()

  instance.onPress()
  expect(pushed).toBe(true)
})

it('should render the detail view', () => {
  const wrapper = shallow(<ColourPickerSelectView />)
  expect(toJSON(wrapper)).toMatchSnapshot()
})

it('calls onChange when a new colour is selected', () => {
  let calledOnChange = false
  let calledPop = false
  GlobalState.navigation.pop = () => { calledPop = true }
  const wrapper = shallow(<ColourPickerSelectView onChange={(value) => {
    expect(value).toBe('#000')
    calledOnChange = true
  }} />)
  const instance = wrapper.instance()
  instance.onPress({ value: '#000', name: 'black' })
  expect(calledOnChange).toBe(true)
  expect(calledPop).toBe(true)
})

it('renders a colour brick', () => {
  const wrapper = shallow(<ColourBrick colour={{ name: 'black', value: '#000' }} />)
  expect(toJSON(wrapper)).toMatchSnapshot()
})

it('renders a light colour brick', () => {
  const wrapper = shallow(<ColourBrick colour={{ name: 'white', value: '#fff' }} />)
  expect(toJSON(wrapper)).toMatchSnapshot()
})

it('handles a colour brick click', () => {
  let didPress = false

  const wrapper = shallow(<ColourBrick colour={{ name: 'white', value: '#fff' }} onPress={(colour) => {
    expect(colour.value).toBe('#fff')
    didPress = true
  }} />)
  const instance = wrapper.instance()

  instance.onPress()
  expect(didPress).toBe(true)
})
