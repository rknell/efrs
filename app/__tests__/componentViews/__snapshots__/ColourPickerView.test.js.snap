// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a colour brick 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  key="black"
  onPress={[Function]}
  style={
    Object {
      "flex": 1,
      "flexBasis": 100,
      "flexGrow": 1,
      "flexShrink": 0,
      "height": 100,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#000",
        "flex": 1,
        "padding": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
        }
      }
    >
      black
    </Text>
  </View>
</TouchableOpacity>
`;

exports[`renders a light colour brick 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  key="white"
  onPress={[Function]}
  style={
    Object {
      "flex": 1,
      "flexBasis": 100,
      "flexGrow": 1,
      "flexShrink": 0,
      "height": 100,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#fff",
        "flex": 1,
        "padding": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "black",
        }
      }
    >
      white
    </Text>
  </View>
</TouchableOpacity>
`;

exports[`should render the detail view 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
    }
  }
>
  <NavBar
    title="Select Colour"
  />
  <ScrollViewMock
    contentContainerStyle={
      Object {
        "flexDirection": "row",
        "flexWrap": "wrap",
        "justifyContent": "flex-start",
      }
    }
    style={
      Object {
        "backgroundColor": "rgba(0,0,0,0.3)",
        "flex": 1,
      }
    }
  >
    <ColourBrick
      colour={
        Object {
          "name": "Red",
          "value": "#f34235",
        }
      }
      key="#f34235"
      onPress={[Function]}
    />
    <ColourBrick
      colour={
        Object {
          "name": "Pink",
          "value": "#e81d62",
        }
      }
      key="#e81d62"
      onPress={[Function]}
    />
    <ColourBrick
      colour={
        Object {
          "name": "Purple",
          "value": "#9b26af",
        }
      }
      key="#9b26af"
      onPress={[Function]}
    />
    <ColourBrick
      colour={
        Object {
          "name": "Indigo",
          "value": "#3e50b4",
        }
      }
      key="#3e50b4"
      onPress={[Function]}
    />
    <ColourBrick
      colour={
        Object {
          "name": "Blue",
          "value": "#2095f2",
        }
      }
      key="#2095f2"
      onPress={[Function]}
    />
    <ColourBrick
      colour={
        Object {
          "name": "Light Blue",
          "value": "#02a8f3",
        }
      }
      key="#02a8f3"
      onPress={[Function]}
    />
    <ColourBrick
      colour={
        Object {
          "name": "Cyan",
          "value": "#00bbd3",
        }
      }
      key="#00bbd3"
      onPress={[Function]}
    />
    <ColourBrick
      colour={
        Object {
          "name": "Teal",
          "value": "#009587",
        }
      }
      key="#009587"
      onPress={[Function]}
    />
    <ColourBrick
      colour={
        Object {
          "name": "Green",
          "value": "#4bae4f",
        }
      }
      key="#4bae4f"
      onPress={[Function]}
    />
    <ColourBrick
      colour={
        Object {
          "name": "Light Green",
          "value": "#8ac249",
        }
      }
      key="#8ac249"
      onPress={[Function]}
    />
    <ColourBrick
      colour={
        Object {
          "name": "Lime",
          "value": "#ccdb38",
        }
      }
      key="#ccdb38"
      onPress={[Function]}
    />
    <ColourBrick
      colour={
        Object {
          "name": "Yellow",
          "value": "#feea3a",
        }
      }
      key="#feea3a"
      onPress={[Function]}
    />
    <ColourBrick
      colour={
        Object {
          "name": "Amber",
          "value": "#fec006",
        }
      }
      key="#fec006"
      onPress={[Function]}
    />
    <ColourBrick
      colour={
        Object {
          "name": "Orange",
          "value": "#fe9700",
        }
      }
      key="#fe9700"
      onPress={[Function]}
    />
    <ColourBrick
      colour={
        Object {
          "name": "Deep Orange",
          "value": "#fe5621",
        }
      }
      key="#fe5621"
      onPress={[Function]}
    />
    <ColourBrick
      colour={
        Object {
          "name": "Brown",
          "value": "#785447",
        }
      }
      key="#785447"
      onPress={[Function]}
    />
    <ColourBrick
      colour={
        Object {
          "name": "Grey",
          "value": "#9d9d9d",
        }
      }
      key="#9d9d9d"
      onPress={[Function]}
    />
    <ColourBrick
      colour={
        Object {
          "name": "Blue Grey",
          "value": "#5f7c8a",
        }
      }
      key="#5f7c8a"
      onPress={[Function]}
    />
    <ColourBrick
      colour={
        Object {
          "name": "Black",
          "value": "#000",
        }
      }
      key="#000"
      onPress={[Function]}
    />
    <ColourBrick
      colour={
        Object {
          "name": "White",
          "value": "#fff",
        }
      }
      key="#fff"
      onPress={[Function]}
    />
  </ScrollViewMock>
</View>
`;

exports[`should render the form control in edit mode 1`] = `
<FieldBase
  editMode={true}
  hidden={false}
  settings={
    Object {
      "test": true,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <TouchableOpacity
      activeOpacity={0.2}
      onPress={[Function]}
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "dodgerblue",
          }
        }
      >
        Pick Colour
      </Text>
    </TouchableOpacity>
  </View>
</FieldBase>
`;

exports[`should render the form control with black selected 1`] = `
<FieldBase
  value="black"
>
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <TouchableOpacity
      activeOpacity={0.2}
      onPress={[Function]}
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "black",
            "padding": 10,
          }
        }
      />
    </TouchableOpacity>
  </View>
</FieldBase>
`;

exports[`should render the form control with no value selected 1`] = `
<FieldBase
  settings={
    Object {
      "test": true,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <TouchableOpacity
      activeOpacity={0.2}
      onPress={[Function]}
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "dodgerblue",
          }
        }
      >
        Pick Colour
      </Text>
    </TouchableOpacity>
  </View>
</FieldBase>
`;
